internal/sqlmigrate/migrate.go

```diff
@@ -0,0 +1,129 @@
+// Package sqlmigrate supports a dbmate-compatible superset of migration files.
+//
+// The superset is that in addition to a migration being a .sql file, it can
+// also be a Go function which is called to execute the migration.
+package sqlmigrate
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io/fs"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/block/ftl/internal/log"
+)
+
+var migrationFileNameRe = regexp.MustCompile(`^.*(\d{14})_(.*)\.sql$`)
+
+type Migration func(ctx context.Context, db *sql.Tx) error
+
+type namedMigration struct {
+	name      string
+	version   string
+	migration string
+}
+
+func (m namedMigration) String() string { return m.name }
+
+// Migrate applies all migrations in the provided fs.FS to the provided database.
+func Migrate(ctx context.Context, dsn *url.URL, migrationFiles fs.FS) error {
+	dsnString := dsn.String()
+	switch dsn.Scheme {
+	case "sqlite", "mysql":
+		dsnString = strings.TrimPrefix(dsnString, dsn.Scheme+"://")
+	}
+	db, err := sql.Open(dsn.Scheme, dsnString)
+	if err != nil {
+		return fmt.Errorf("failed to open database: %w", err)
+	}
+	defer db.Close() //nolint:errcheck
+
+	// Create schema_migrations table if it doesn't exist.
+	// This table structure is compatible with dbmate.
+	_, _ = db.ExecContext(ctx, `CREATE TABLE schema_migrations (version TEXT PRIMARY KEY)`) //nolint:errcheck
```

Comment:
If we add a digest column here we could detect when the migration has failed for local dev.

Reply:
That is a great idea. When we revisit we should do that.